32662 enable align content (#32862)
mdn/content: files/en-us/web/css/align-content/index.md

{{CSSRef}}

> The [CSS](/en-US/docs/Web/CSS) **`align-content`** property sets the distribution of space between and around content items along a [flexbox](/en-US/docs/Web/CSS/CSS_flexible_box_layout)'s [cross-axis](/en-US/docs/Glossary/Cross_Axis), a [grid](/en-US/docs/Web/CSS/CSS_grid_layout)'s or [block](/en-US/docs/Glossary/Block-level_content)'s block axis.

Reviewer: We need to add block, and possibly multi-col to BCD

- `unsafe`
  - : Used alongside an alignment keyword. Regardless of the relative sizes of the item and alignment container and whether overflow which causes data loss might happen, the given alignment value is honored.

> **Note:** The `<content-distribution>` values (`space-between`, `space-around`, `space-evenly`, and `stretch`) have no effect in [block layout](/en-US/docs/Web/CSS/CSS_box_alignment/Box_alignment_in_block_abspos_tables#align-content_and_justify-content) as all the content in that block is treated as a single [alignment-subject](/en-US/docs/Glossary/Alignment_Subject)

> ## Formal definition

Author: This is referenced in the data repo https://github.com/mdn/data/pull/721/files which I have updated and waiting for this to be approved.

## Examples

> ### align-content switcher

Reviewer: ```suggestion
```
Only one example, so we can use "examples" in the Embed live sample, and not have this title

> In this example you can switch between the different layouts flex, grid, and block. You can also switch between the different values for `align-content`.

Reviewer: ```suggestion
In this example, you can switch between three different {{cssxref("display")}} property values, includ `flex`, `grid`, and `block`. You can also switch between the different values for `align-content`.
```

#### HTML

```html-nolint hidden
<div class="wrapper">
```

```html

> <section id="outer" class="block">
>   <div id="inner" class="olive"></div>
>   <div id="inner" class="coral"></div>
>   <div id="inner" class="deepskyblue"></div>
> </section>

Reviewer: ```suggestion
<section>
  <div class="olive"></div>
  <div class="coral"></div>
  <div class="deepskyblue"></div>
</section>
```

```

```html hidden

> <section class="controls">

Reviewer: ```suggestion
<fieldset class="controls">
```

>     <h2>Controls</h2>

Reviewer: ```suggestion
    <legend>Controls</legend>
```

    <div class="row">
      <label for="display">display: </label>
      <select id="display">
        <option value="flex">flex</option>
        <option value="grid">grid</option>
        <option value="block">block</option>
      </select>
    </div>
    <div class="row">
      <label for="values">align-content: </label>
      <select id="values">
        <option value="normal" selected>normal</option>
        <option value="start">start</option>
        <option value="center">center</option>
        <option value="end">end</option>
        <option value="flex-start">flex-start</option>
        <option value="flex-end">flex-end</option>
        <option value="space-between">space-between</option>
        <option value="space-around">space-around</option>
        <option value="space-evenly">space-evenly</option>
      </select>
    </div>
    <h2>styles</h2>
    <pre>
      .<span class="displayType">flex</span> {
        display: <span class="displayType">flex</span>;
        align-content: <span id="align">normal</span>
      }
    </pre>

>     <p id="note" hidden="hidden">
>       In <a href="/en-US/docs/Web/CSS/CSS_box_alignment/Box_alignment_in_block_abspos_tables#align-content_and_justify-content">block layout the child elements are treated a single element</a><br>so  <code class="align">normal</code> behaves differently.
>     </p>

Reviewer: ```suggestion
```

>   </section>

Reviewer: ```suggestion
  </fieldset>
```

</div>
```

#### CSS

```css hidden
.wrapper {
  display: flex;
  gap: 1rem;
}
h2 {
  margin: 0;
  margin-bottom: 0.2rem;
}
.row {
  margin-bottom: 0.2rem;
}
```

```css

> #outer {

Reviewer: ```suggestion
section {
```

  border: solid 2px tomato;
  height: 300px;
  width: 300px;
}

> #inner {

Reviewer: ```suggestion
section div {
```

  height: 75px;
  width: 100%;
}
.olive {
  background-color: olive;
}
.coral {
  background-color: coral;
}
.deepskyblue {
  background-color: deepskyblue;
}
.flex {
  display: flex;
  flex-wrap: wrap;
  gap: 0.2rem;
}
.grid {
  display: grid;
  gap: 0.2rem;
}
.block {
  display: block;
}
```

```js hidden
const values = document.querySelector("#values");
const display = document.querySelector("#display");

> const container = document.querySelector("#outer");

Reviewer: ```suggestion
const container = document.querySelector("section");
```

const displayTypes = document.querySelectorAll(".displayType");
const aligns = document.querySelectorAll(".align");

> const note = document.querySelector("#note");
> const hideNote = (a, d) => {
>   const distributions = ["space-between", "space-around", "space-evenly"];
>   if (distributions.includes(a) && d === "block") {
>     note.removeAttribute("hidden");
>   } else {
>     note.setAttribute("hidden", "hidden");
>   }
> };

Reviewer: ```suggestion
```
Let's make the note alwasy visible under the results as people should be told this and it should be searchable

values.addEventListener("change", (evt) => {
  const alVal = evt.target.value;
  container.style.alignContent = alVal;
  aligns.forEach((align) => {
    align.innerText = alVal;
  });

>   hideNote(alVal, display.value);

Reviewer: ```suggestion
```

});
display.addEventListener("change", (evt) => {
  const dVal = evt.target.value;
  container.className = dVal;
  displayTypes.forEach((dType) => {
    dType.innerText = dVal;
  });

>   hideNote(values.value, dVal);

Reviewer: ```suggestion
```

});
```

#### Result

Try changing the `display` value and the `align-content` value.

> {{EmbedLiveSample("align-content_switcher", 260, 310)}}

Reviewer: ```suggestion
{{EmbedLiveSample("Examples", 260, 310)}}

In [block layout](en-US/docs/Web/CSS/CSS_box_alignment/Box_alignment_in_block_abspos_tables#align-content_and_justify-content), child elements are treated as a single element, meaning `normal` behaves differently.
```

</section>
```

```html-nolint hidden
<fieldset class="controls">
    <legend>Controls</legend>
    <div class="row">
      <label for="display">display: </label>
      <select id="display">
        <option value="block" selected>block</option>
        <option value="flex">flex</option>
        <option value="grid">grid</option>
      </select>
    </div>
    <div class="row">
      <label for="alignContent">align-content: </label>
      <select id="alignContent">
        <option value="normal" selected>normal</option>
        <option value="start">start</option>
        <option value="center">center</option>
        <option value="end">end</option>
        <option value="flex-start">flex-start</option>
        <option value="flex-end">flex-end</option>
        <option value="space-between">space-between</option>
        <option value="space-around">space-around</option>
        <option value="space-evenly">space-evenly</option>
      </select>
    </div>
    <p>CSS applied</p>
    <pre>
section {
  display: <span id="displayStyle">block</span>;
  align-content: <span id="align">normal</span>
}
    </pre>
  </fieldset>
</div>
```

#### CSS

```css hidden

> @layer pageSetUp {

Reviewer: the layer in the codepen was my way of saying "hidden". you can keep it, but it may confuse people so i would delete this line and the closing on line 172

Author: Done

  .wrapper {
    font-size: 1.25rem;
    display: flex;
    gap: 1rem;
  }
  section div {
    font-family: monospace;
    padding: 3px;
  }
}
```

```css
section {
  border: solid 1.5px tomato;
  height: 300px;
  width: 300px;
  flex-wrap: wrap; /* used by flex only */
  gap: 0.2rem;
}
/* section div {
  height: 75px;
  width: 100%;

> } */

Reviewer: these were options to play with. delete, but suggesting adding these lines in the "results" section might be good

Author: I just removed this as, to me, it doesn't make sense when having the `flex-wrap` property.

.olive {
  background-color: olive;
}
.coral {
  background-color: coral;
}
.deepskyblue {
  background-color: deepskyblue;
}
.orchid {
  background-color: orchid;
}
.slateblue {
  background-color: slateblue;
  color: white;
}
.maroon {
  background-color: maroon;
  color: white;

> }
> .flex {
>   display: flex;
>   flex-wrap: wrap;
>   gap: 0.2rem;
> }
> .grid {
>   display: grid;
>   gap: 0.2rem;
> }
> .block {
>   display: block;

Reviewer: ```suggestion
```
we're using `js.style`, rather than class changes, so this CSS isn't used

}
```

```js hidden
const alignContent = document.querySelector("#alignContent");
const display = document.querySelector("#display");
const container = document.querySelector("section");
const displayStyle = document.querySelector("#displayStyle");
const alignStyle = document.querySelector("#align");
document.addEventListener("load", () => {
  updatePage();
});
alignContent.addEventListener("change", () => {
  updatePage();
});
display.addEventListener("change", () => {
  updatePage();
});
function updatePage() {
  const alVal = alignContent.value;
  const dVal = display.value;
  container.style.alignContent = alVal;
  container.style.display = dVal;
  alignStyle.innerText = alVal;
  displayStyle.innerText = dVal;
}
```

#### Result

Try changing the `display` value and the `align-content` value.

{{EmbedLiveSample("Examples", 260, 310)}}

> In [block layout](/en-US/docs/Web/CSS/CSS_box_alignment/Box_alignment_in_block_abspos_tables#align-content_and_justify-content), child elements are treated as a single element, meaning `space-between`, `space-around`, and `space-evenly` behaves differently.

Reviewer: ```suggestion
In [block layout](/en-US/docs/Web/CSS/CSS_box_alignment/Box_alignment_in_block_abspos_tables#align-content_and_justify-content), child elements are treated as a single element, meaning `space-between`, `space-around`, and `space-evenly` behave differently.
```
